doc/developer/design/20230607_shift_status_updates.md

```diff
@@ -0,0 +1,150 @@
+
+## Goals
+
+<!--
+Enumerate the concrete goals that are in scope for the project.
+-->
+
+- Describe how to shift the writing of status updates for sources/sinks from storage workers to the storage controller
+- How to evolve status relations to support storage clusters with multiple replicas
+
+<!-- ## Non-Goals -->
+
+<!--
+Enumerate potential goals that are explicitly _out_ of scope for the project
+ie. what could we do or what do we want to do in the future - but are not doing
+now.
+-->
+
+## Overview
+
+<!--
+Brief, high-level overview. A few sentences long, at most a couple of smaller
+paragraphs.
+-->
+
+The primary motivation for enforcing status updates flow through the storage controller is that it allows us to correctly handle instances in which a replica is dropped, in which case any late status updates can simply be ignored and the status of any dataflows managed by the dropped replica can have their status set to `paused`. As mentioned in the [Alternatives](#alternatives) section, relying on graceful shutdown instead would be complex and error prone.
+
+The health operator for ingestion dataflows will continue to exist as a terminal operator in the dataflow graph. However, rather than writing status updates directly to the persist shard corresponding to the collection's status history, the health operator will update a local mapping from `GlobalId` to `HealthStatus` that exists as part of the storage worker's state. Updates of a similar fashion will be done during sink production as well.
+
+Upon some defined frequency, storage workers will emit a new `StorageResponse::StatusUpdates` to the storage controller. When receiving these events, the storage controller will write the status updates to the shard associated with `mz_source_status_history` / `mz_sink_status_history`.
+
+## Detailed description
+
+<!--
+Describe the approach in detail. If there is no clear frontrunner, feel free to
+list all approaches in alternatives. If applicable, be sure to call out any new
+testing/validation that will be required.
+
+For some features it can be helpful to sketch an implementation. If you're
+working on things that are crossing team boundaries it will be helpful to spell
+out any new interfaces/traits/interactions.
+
+For most new features, you should think about testing, rollout/lifecycle, and
+observability. These things can warrant their own sections.
+-->
+
+We will introduce a couple new pieces of state to `mz_storage::storage_state::StorageState` in order to keep track of the statuses of sink and source ingestion dataflows:
+
+```rust
+pub source_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
+pub sink_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
```

Comment:
```suggestion
pub source_statuses: Rc<RefCell<BTreeMap<GlobalId, Vec<HealthStatus>>>>,
pub sink_statuses: Rc<RefCell<BTreeMap<GlobalId, Vec<HealthStatus>>>>,
```

Wondering if these ought to be `Vec`s rather than `Option`s. Seems like the storage controller might want to deliver the _sequence_ of states that the source/sink transitioned through even if the transitions were to fast to be noticed by the poll loop.

Actually, I wonder if even better would be a single `object_statuses` channel of type `ObjectStatusUpdate`. Then every source/sink implementation could just fire its updates into the channel, and the main processing loop could fire any pending updates ASAP to the storage controller. That's all cheap. The only expensive part is actually writing to persist, and we could do that throttling in the storage controller.

Reply:
Definitely agree on a single channel, I think it simplifies the code a ton.

My initial thinking was to perform some throttling in the storage worker as an attempt to preserve network throughput, especially if writes to persist are going to happen in larger intervals on the controller. But will defer to your judgement on whether that makes much of a difference. If not happy to send any available updates each tick.

Reply:
> Seems like the storage controller might want to deliver the sequence of states that the source/sink transitioned through even if the transitions were to fast to be noticed by the poll loop.

If we care about accurately capturing the sequence, we may need to modify / elaborate on the bit below that says status is flushed only periodically to the controller.

My personal opinion is that we mostly care about either the latest status or longer-term trends, and some debouncing is fine and healthy. But that certainly doesn't feel like the only reasonable opinion. :)

Reply:
> If we care about accurately capturing the sequence, we may need to modify / elaborate on the bit below that says status is flushed only periodically to the controller.

Yeah, I think in this case we'd want to capture timestamps in the status updates themselves. That way we can still batch writes and/or flush updates periodically to the controller. The currently suggested approach would just set the timestamp of the latest update in the controller immediately before writing to persist.

I do think that from a user perspective, if we're throttling status updates in the poll loop it makes sense to only send the latest status update. If we were to deliver the sequence of updates, it might be unnecessarily granular given that any further computation on the status rows currently just takes into account the latest update.

Reply:
I still think there's room to consolidate into one data structure for source/sink status updates, regardless of whether that's a map or channel. But will wait until we reach a consensus here before updating the doc!

Reply:
> Yeah, I think in this case we'd want to capture timestamps in the status updates themselves.

This sounds like a great approach.

> If we care about accurately capturing the sequence, we may need to modify / elaborate on the bit below that says status is flushed only periodically to the controller.

Yeah, I think it's totally okay for the status to be a bit stale (5s, 10s, a minute, whatever) but it has proven to be confusing when we coalesce state transitions. In the past we've had to explain to people that seeing lots of running status in a row...

```
1:05:00pm running
1:05:30pm running
1:06:00pm running
```

...indicates a crash loop. Which has happened in the past, IIRC, in some source implementation that wasn't properly emitting a "started" status. It's much more obvious that something is wrong when the "starting" status is interleaved:

```
1:05:00pm starting
1:05:01pm running
1:05:31pm starting
1:05:31pm running
1:06:00pm starting
1:06:01pm running
```